script/bulk_import/generic_bulk.rb

```diff
@@ -2006,6 +2007,66 @@ def import_votes
     puts "  Update took #{(Time.now - start_time).to_i} seconds."
   end
 
+  def import_topic_voting_votes
+    unless defined?(::DiscourseTopicVoting)
+      puts "", "Skipping topic voting votes, because the topic voting plugin is not installed."
+      return
+    end
+
+    puts "", "Importing votes for topics..."
+
+    topic_votes = query(<<~SQL)
+      SELECT *
+      FROM topic_voting_votes
+    SQL
+
+    existing_topic_votes = DiscourseTopicVoting::Vote.pluck(:topic_id, :user_id).to_set
+
+    create_topic_voting_votes(topic_votes) do |row|
+      topic_id = topic_id_from_imported_id(row["topic_id"])
+      user_id = user_id_from_imported_id(row["user_id"])
+
+      next unless topic_id && user_id
+      next unless existing_topic_votes.add?([topic_id, user_id])
+
+      {
+        topic_id: topic_id,
+        user_id: user_id,
+        archive: to_boolean(row["archive"]),
+        created_at: to_datetime(row["created_at"]),
+        updated_at: to_datetime(row["updated_at"]),
+      }
+    end
+
+    topic_votes.close
+
+    puts "", "Updating vote counts of topics..."
+
+    start_time = Time.now
+
+    DB.exec(<<~SQL)
+      WITH missing_ids AS (
+        SELECT DISTINCT t.id FROM topics t
+        JOIN topic_voting_votes dvv ON t.id = dvv.topic_id
+        LEFT JOIN topic_voting_topic_vote_count dvtvc ON t.id = dvtvc.topic_id
+        WHERE dvtvc.topic_id IS NULL
+      )
+      INSERT INTO topic_voting_topic_vote_count (votes_count, topic_id, created_at, updated_at)
+      SELECT '0', id, now(), now() FROM missing_ids
+    SQL
+
+    DB.exec(<<~SQL)
+      UPDATE topic_voting_topic_vote_count dvtvc
+      SET votes_count = (
+        SELECT COUNT(*) FROM topic_voting_votes dvv
+        WHERE dvtvc.topic_id = dvv.topic_id
+        GROUP BY dvv.topic_id
+      )
+    SQL
```

Comment:
The current logic inserts default values into the `topic_voting_topic_vote_count` table, and then it updates all rows with the calculated count. I'd try to make this a bit faster, by running the `UPDATE` first, and only then insert missing records. That way the DB doesn't have to write to the same rows twice.

Untested code, but this is what I'd do:

```suggestion
    DB.exec(<<~SQL)
        WITH
          votes AS (
                     SELECT topic_id, COUNT(*) AS votes_count
                     FROM topic_voting_votes
                     GROUP BY topic_id
                   )
      UPDATE topic_voting_topic_vote_count
         SET votes_count = votes.votes_count
        FROM votes
       WHERE votes.topic_id = topic_voting_topic_vote_count.topic_id
         AND votes.votes_count <> topic_voting_topic_vote_count.votes_count
    SQL

    DB.exec(<<~SQL)
        WITH
          missing_votes AS (
                             SELECT v.topic_id, COUNT(*) AS votes_count
                               FROM topic_voting_votes v
                              WHERE NOT EXISTS (
                                                 SELECT 1
                                                 FROM topic_voting_topic_vote_count c
                                                 WHERE v.topic_id = c.topic_id
                                               )
                              GROUP BY topic_id
                           )
      INSERT
        INTO topic_voting_topic_vote_count (votes_count, topic_id, created_at, updated_at)
      SELECT votes_count, topic_id, NOW(), NOW()
        FROM missing_votes
    SQL
```

A couple of notes:
* The `UPDATE` contains `votes.votes_count <> topic_voting_topic_vote_count.votes_count` as a condition to prevent unnecessary writes.
* The `missing_votes` CTE calculates the `votes_count` for all topics in `topic_voting_votes` that are missing from `topic_voting_topic_vote_count`.
* I prefer writing `EXISTS` / `NOT EXISTS` subqueries. They are a bit more verbose, but I think it's easier to understand what's happening. But of course, your solution with `LEFT JOIN` & `IS NULL` works as well.

Reply:
Hey @gschlager, I think these new queries are great! Perhaps we should update the [Topic Votes plugin queries](https://github.com/discourse/discourse-topic-voting/blob/a39bb09bc9d17d1dc77084f20eb0ddd77c97b9a9/db/post_migrate/20240711102255_ensure_consistency.rb#L43-L63) as well? That's where I got them from.

Reply:
I'm reluctant to change those queries as they are in a DB migration that executes only once, and most sites probably have already executed them anyway. But I made a mental note to take a closer look at that plugin some time. I think it has duplicate indexes and possibly other DB problems.